File: 45-jump-game-ii/jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        if (n <= 1) return 0;

        int j = 0;
        int curr = 0;
        int maxval = 0;

        for (int i = 0; i < n - 1; i++) {
            maxval = Math.max(maxval, i + nums[i]);

            if (i == curr) {
                j++;
                curr = maxval;

                if (curr >= n - 1) break;
            }
        }

        return j;
    }
}

File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
       int n= nums.size();
       if(n==1){
           return 0;
       } 
       int jumps=0;
       int curr_end=0;
       int curr_far=0;
       for(int i=0;i<n-1;++i){
           curr_far=max(curr_far,i+nums[i]);
           if(i==curr_end){
               curr_end =curr_far;
               jumps++;
           }
       }
       return jumps;
    }
};

